### PR TITLE
Fix lost combinator before global

### LINKED ISSUE #755: bug: child selector `>` disappears with `:global()`
#### Do you want to request a *feature* or report a *bug*?

bug

#### What is the current behavior?

`div > :global(span)` becomes `div.jsx-c936d37f15b46a3d span`

#### If the current behavior is a bug, please provide the steps to reproduce and possibly a minimal demo or testcase in the form of a Next.js app, CodeSandbox URL or similar

https://codesandbox.io/s/friendly-meitner-op7qg?file=/pages/index.js

#### What is the expected behavior?

`div > :global(span)` becomes `div.jsx-c936d37f15b46a3d > span`

#### Environment (include versions)

OS: codesandbox or linux
Browser: chrome, not that i believe it matters
styled-jsx (version): buggy in 3.3.0 and 5.0.0-beta.3 too

#### Did this work in previous versions?

It was working on next11 with babel.

### LINKED ISSUE #30390: CSS Direct Child Selector with :global() no longer works in styled-jsx
### What version of Next.js are you using?

12.0.0

### What version of Node.js are you using?

v16.11.0

### What browser are you using?

Chrome

### What operating system are you using?

Windows

### How are you deploying your application?

Locally (next dev)

### Describe the Bug

It seems like direct child CSS combinators (x > y) is not working, at least with the presence of `:global()` selectors.

(See Repro for example)

The output CSS will look like this:

```css
.wm.jsx-XYZ *:not(.fullwidth):not(.wider) {
  /* ... */
}
```

rather than ">".

### Expected Behavior

Output CSS will look like:

```css
.wm.jsx-XYZ>*:not(.fullwidth):not(.wider) {
  /* ... */
}
```

which is what works in v11.

### To Reproduce

Take CSS Like this:

```tsx
const wmStyle = css.resolve`
  .wm > :global(*:not(.fullwidth):not(.wider)) {
    max-width: 5rem;
`;
const Component = () => (
  <>
    <div className={wmStyle.className}>
      <div id="first">
        <div id="second"></div>
      </div>
    </div>
    {wmStyle.styles}
  </>
);
```

The output CSS will look like this:

```css
.wm.jsx-XYZ *:not(.fullwidth):not(.wider) {
  /* ... */
}
```

rather than ">".

### PR DESCRIPTION
**Pull request by padmaia**
Fixes https://github.com/vercel/styled-jsx/issues/755
Fixes https://github.com/vercel/next.js/issues/30390

## Bug

- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Errors have helpful link attached, see `contributing.md`

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [ ] Make sure the linting passes by running `yarn lint`

### CODE REVIEW

**Review by ijjk — DISMISSED** [review]

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| buildDuration | 21.6s | 21.4s | -211ms |
| buildDurationCached | 4.5s | 4.5s | -45ms |
| nodeModulesSize | 293 MB | 293 MB | ⚠️ +12 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.457 | 3.514 | ⚠️ +0.06 |
| / avg req/sec | 723.23 | 711.39 | ⚠️ -11.84 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.549 | 1.597 | ⚠️ +0.05 |
| /error-in-render avg req/sec | 1613.71 | 1565.76 | ⚠️ -47.95 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 28 kB | 28 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 71.9 kB | 71.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| polyfills-a4..dd70.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.23 kB | 1.23 kB | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 2.38 kB | 2.38 kB | ✓ |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 635 B | 635 B | ✓ |
| image-HASH.js gzip | 4.44 kB | 4.44 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 1.87 kB | 1.87 kB | ✓ |
| routerDirect..HASH.js gzip | 321 B | 321 B | ✓ |
| script-HASH.js gzip | 383 B | 383 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 334f979574ae..6f4.css gzip | 106 B | 106 B | ✓ |
| Overall change | 13.1 kB | 13.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| index.html gzip | 533 B | 533 B | ✓ |
| link.html gzip | 546 B | 546 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.61 kB | 1.61 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| buildDuration | 23.3s | 23.4s | ⚠️ +53ms |
| buildDurationCached | 4.7s | 4.4s | -232ms |
| nodeModulesSize | 293 MB | 293 MB | ⚠️ +12 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.389 | 3.406 | ⚠️ +0.02 |
| / avg req/sec | 737.73 | 733.91 | ⚠️ -3.82 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.687 | 1.592 | -0.09 |
| /error-in-render avg req/sec | 1481.95 | 1570.73 | +88.78 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.2 kB | 28.2 kB | ✓ |
| webpack-HASH.js gzip | 1.43 kB | 1.43 kB | ✓ |
| Overall change | 72.1 kB | 72.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| polyfills-a4..dd70.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.22 kB | 1.22 kB | ✓ |
| _error-HASH.js gzip | 180 B | 180 B | ✓ |
| amp-HASH.js gzip | 305 B | 305 B | ✓ |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.38 kB | 2.38 kB | ✓ |
| head-HASH.js gzip | 342 B | 342 B | ✓ |
| hooks-HASH.js gzip | 622 B | 622 B | ✓ |
| image-HASH.js gzip | 4.46 kB | 4.46 kB | ✓ |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 1.91 kB | 1.91 kB | ✓ |
| routerDirect..HASH.js gzip | 314 B | 314 B | ✓ |
| script-HASH.js gzip | 375 B | 375 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 334f979574ae..6f4.css gzip | 106 B | 106 B | ✓ |
| Overall change | 13.1 kB | 13.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| _buildManifest.js gzip | 460 B | 460 B | ✓ |
| Overall change | 460 B | 460 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| index.html gzip | 534 B | 534 B | ✓ |
| link.html gzip | 547 B | 547 B | ✓ |
| withRouter.html gzip | 528 B | 528 B | ✓ |
| Overall change | 1.61 kB | 1.61 kB | ✓ |

</details>

</details>
Commit: ff0657d217c1acc63557658e1835fb65a64814a0

File: packages/next/build/swc/src/styled_jsx/transform_css.rs [hunk 1]
```diff
@@ -187,6 +187,10 @@ impl Namespacer {
                                 .cloned()
                                 .collect::<Vec<_>>();
 
+                            if node.combinator.is_some() {
+                                v[0].combinator = node.combinator;
```
Comment (kdy1) [inline packages/next/build/swc/src/styled_jsx/transform_css.rs:191]: Can be `v[0].combinator` some?
If so, it can silently swallow an error.

**Review by kdy1 — COMMENTED** [review]
I'm not sure about if it's possible to make such input

CSS like below is invalid according to w3c css validator, so I'm not sure if it can be actual input.
```css
div > :global( > span) {
    color: red;
}
```

File: packages/next/build/swc/src/styled_jsx/transform_css.rs (hunk 1 above)
Comment (kdy1) [reply packages/next/build/swc/src/styled_jsx/transform_css.rs:191]: Thanks! Then it's fine.

**Comment by padmaia** [comment]
Looks like the Babel plugin will output invalid CSS `div.jsx-3020906210>> span{color:red;}`. The swc transform will fail with an error now.

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| buildDuration | 21.6s | 21.7s | ⚠️ +12ms |
| buildDurationCached | 4.6s | 4.6s | -9ms |
| nodeModulesSize | 293 MB | 293 MB | -8 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.39 | 3.418 | ⚠️ +0.03 |
| / avg req/sec | 737.37 | 731.51 | ⚠️ -5.86 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.632 | 1.64 | ⚠️ +0.01 |
| /error-in-render avg req/sec | 1531.81 | 1524.06 | ⚠️ -7.75 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 28 kB | 28 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 71.9 kB | 71.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.23 kB | 1.23 kB | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 2.38 kB | 2.38 kB | ✓ |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 635 B | 635 B | ✓ |
| image-HASH.js gzip | 4.44 kB | 4.44 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 1.87 kB | 1.87 kB | ✓ |
| routerDirect..HASH.js gzip | 321 B | 321 B | ✓ |
| script-HASH.js gzip | 383 B | 383 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 334f979574ae..6f4.css gzip | 106 B | 106 B | ✓ |
| Overall change | 13.1 kB | 13.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| index.html gzip | 521 B | 521 B | ✓ |
| link.html gzip | 533 B | 533 B | ✓ |
| withRouter.html gzip | 515 B | 515 B | ✓ |
| Overall change | 1.57 kB | 1.57 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| buildDuration | 23.2s | 23.2s | -18ms |
| buildDurationCached | 4.6s | 4.6s | -25ms |
| nodeModulesSize | 293 MB | 293 MB | -8 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.363 | 3.542 | ⚠️ +0.18 |
| / avg req/sec | 743.34 | 705.89 | ⚠️ -37.45 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.573 | 1.763 | ⚠️ +0.19 |
| /error-in-render avg req/sec | 1588.94 | 1418.25 | ⚠️ -170.69 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.2 kB | 28.2 kB | ✓ |
| webpack-HASH.js gzip | 1.43 kB | 1.43 kB | ✓ |
| Overall change | 72.1 kB | 72.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.22 kB | 1.22 kB | ✓ |
| _error-HASH.js gzip | 180 B | 180 B | ✓ |
| amp-HASH.js gzip | 305 B | 305 B | ✓ |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.38 kB | 2.38 kB | ✓ |
| head-HASH.js gzip | 342 B | 342 B | ✓ |
| hooks-HASH.js gzip | 622 B | 622 B | ✓ |
| image-HASH.js gzip | 4.46 kB | 4.46 kB | ✓ |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 1.91 kB | 1.91 kB | ✓ |
| routerDirect..HASH.js gzip | 314 B | 314 B | ✓ |
| script-HASH.js gzip | 375 B | 375 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 334f979574ae..6f4.css gzip | 106 B | 106 B | ✓ |
| Overall change | 13.1 kB | 13.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| _buildManifest.js gzip | 460 B | 460 B | ✓ |
| Overall change | 460 B | 460 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | padmaia/next.js fix/child-selector-before-global | Change |
| - | - | - | - |
| index.html gzip | 521 B | 521 B | ✓ |
| link.html gzip | 534 B | 534 B | ✓ |
| withRouter.html gzip | 516 B | 516 B | ✓ |
| Overall change | 1.57 kB | 1.57 kB | ✓ |

</details>

</details>
Commit: 9e4694a946da327bccc6deeec05ec9e8764403d4

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 9e4694a946da327bccc6deeec05ec9e8764403d4

**test/integration/preload-viewport/test/index.test.js** 
- Prefetching Links in viewport > should not prefetch already loaded scripts

<details>
<summary>Expand output</summary>

  ● Prefetching Links in viewport › should not prefetch already loaded scripts

    expect(received).toBe(expected) // Object.is equality

    Expected: false
    Received: true

      397 |     expect(hrefs).toEqual([...new Set(hrefs)])
      398 |
    > 399 |     // Verify encoding
          |     ^
      400 |     expect(hrefs.some((e) => e.includes(`%5Bhello%5D-`))).toBe(true)
      401 |   })
      402 |

      at Object.<anonymous> (integration/preload-viewport/test/index.test.js:399:5)

</details>

**Review by kdy1 — APPROVED** [review]
I checked the code to know if `v` being empty is possible for normal inputs, but it does not seem like a valid input.
So I think this change is fine.

**Pull request merged**